### ISSUE TITLE
Function declaration in condition branches reported to be obscured

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1413]
**Describe the bug**
When a fuction is declared conditionally in multiple branches, each preceding declaration is reported to be obscured.  This is a regression introduced in 1.1.104.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That it wouldn't produce any type errors.

**Screenshots or Code**
```python
# pyright: strict

if 1:
    # Function declaration "foo" is obscured by a declaration of the same name
    def foo(): ...
else:
    def foo(): ...
```

**VS Code extension or command-line**
Either, 1.1.104

**Comment by erictraut** [comment]
This behavior is intended.

**Comment by layday** [comment]
Well, there is definitely value in reporting function redaclarations and I don't fully expect Pyright to distinguish between redeclarations in conditionals which happen to be in the same typing context because of Python variable scoping rules and actual overwrites.   However, the alternative is a little unsightly and I'd have no reason to re-assign a function to a variable other than to placate the type checker.  I'm not sure what to suggest.

**Comment by erictraut** [comment]
I'll leave it to you to decide whether the type checker is providing sufficient benefit to you in other ways to justify making minor changes to "placate" it.

**Comment by layday** [comment]
Well, yes, it is, and I didn't mean to imply that Pyright is even at fault; just a poor choice of words.
